dbcrossbarlib/src/drivers/postgres/csv_to_binary/mod.rs

```diff
@@ -0,0 +1,331 @@
+        PgScalarDataType::Bigint => write_json_as_binary::<i64, W>(wtr, json),
+        PgScalarDataType::Json => Err(format_err!(
+            "PostgreSQL arrays with json elements not supported (try jsonb)",
+        )),
+        PgScalarDataType::Jsonb => {
+            let serialized = serde_json::to_string(json)?;
+            RawJsonb(&serialized).write_binary(wtr)
+        }
+        PgScalarDataType::Text => match json {
+            Value::String(s) => s.as_str().write_binary(wtr),
+            _ => Err(format_err!("expected JSON string, found {}", json)),
+        },
+        PgScalarDataType::TimestampWithoutTimeZone => {
+            write_json_as_binary::<NaiveDateTime, W>(wtr, json)
+        }
+        PgScalarDataType::TimestampWithTimeZone => {
+            write_json_as_binary::<DateTime<Utc>, W>(wtr, json)
+        }
+        PgScalarDataType::Uuid => write_json_as_binary::<Uuid, W>(wtr, json),
+    }
+}
+
+/// Parse a JSON value and write it out as a PostgreSQL binary value. This works
+/// for any type implementing `FromJsonValue` and `WriteBinary`. More
+/// complicated cases will need to do this manually.
+fn write_json_as_binary<T, W>(wtr: &mut W, json: &Value) -> Result<()>
+where
+    T: FromJsonValue + WriteBinary,
+    W: Write,
+{
+    let value = T::from_json_value(json)?;
+    value.write_binary(wtr)
+}
+
+/// Convert a scalar value from a CSV file into a `BINARY` value.
+fn scalar_to_binary(
+    wtr: &mut BufferedWriter,
+    data_type: &PgScalarDataType,
+    cell: &str,
+) -> Result<()> {
+    match data_type {
+        PgScalarDataType::Boolean => write_cell_as_binary::<bool>(wtr, cell),
+        PgScalarDataType::Date => write_cell_as_binary::<NaiveDate>(wtr, cell),
+        PgScalarDataType::Numeric => {
+            // The only sensible way to make this work is to port PostgresSQL's
+            // own `decimal` parser from C, because it's an unusual internal
+            // format built using very complicated parsing rules (and `numeric`
+            // needs to be a perfectly-accurate type).
+            Err(format_err!(
+                "cannot use numeric columns with PostgreSQL yet",
+            ))
+        }
+        PgScalarDataType::Real => write_cell_as_binary::<f32>(wtr, cell),
+        PgScalarDataType::DoublePrecision => write_cell_as_binary::<f64>(wtr, cell),
+        PgScalarDataType::Geometry(srid) => {
+            let geometry = Geometry::<f64>::from_csv_cell(cell)?;
+            let value = GeometryWithSrid {
+                geometry: &geometry,
+                srid: *srid,
+            };
+            value.write_binary(wtr)
+        }
+        PgScalarDataType::Smallint => write_cell_as_binary::<i16>(wtr, cell),
+        PgScalarDataType::Int => write_cell_as_binary::<i32>(wtr, cell),
+        PgScalarDataType::Bigint => write_cell_as_binary::<i64>(wtr, cell),
+        PgScalarDataType::Json => Err(format_err!(
+            "PostgreSQL json columns not supported (try jsonb)",
```

Comment:
@seamusabshere It's actually pretty trivial at the driver level. The actual obstacle is the "idealized portable schema" level, where we only have one `Json` type, that maps to PostgreSQL `jsonb`, because nobody should be using `json` if I understand correctly.

Reply:
@emk ok but - do we support sending data to and from the legacy json? because people (including us) definitely do have json cols that are hard to transition to jsonb because of legacy function usage

Reply:
At the moment, no. It's not completely impossible to support this, but it would require some extra steps and better database specific schema processing.

We have no problem copying _out_ of `json` columns, in theory. It's creating new ones or appending to tables with them that we don't do.